File: 21-merge-two-sorted-lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(-1);

        ListNode temp = dummy;

        ListNode curr1 = list1, curr2 = list2;

        while(curr1 != null || curr2 != null){
            if(curr1 != null && curr2 != null){
                if(curr1.val >= curr2.val){
                    temp.next = curr2;
                    curr2 = curr2.next;
                    temp = temp.next;
                }
                else{
                    temp.next = curr1;
                    curr1 = curr1.next;
                    temp = temp.next;
                }
            }
            else if(curr1 != null){
                temp.next = curr1;
                curr1 = curr1.next;
                temp = temp.next;
            }
            else{
                temp.next = curr2;
                curr2 = curr2.next;
                temp = temp.next;
            }
        }


        return dummy.next;
    }
}

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (!list1) return list2;
        if (!list2) return list1;

        ListNode* dummy = new ListNode();
        ListNode* current = dummy;

        while (list1 && list2) {
            if (list1->val <= list2->val) {
                current->next = list1;
                list1 = list1->next;
            } else {
                current->next = list2;
                list2 = list2->next;
            }
            current = current->next;
        }

        if (list1) current->next = list1;
        if (list2) current->next = list2;

        return dummy->next;
    }
};

/*
<-- RECURSION -->

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (!list1) {
            return list2;
        }
        if (!list2) {
            return list1;
        }
        if (list1->val <= list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        } 
        else {
            list2->next = mergeTwoLists(list1, list2->next);
            return list2;
        }
    }
};

<-- BRUTE FORCE -->

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        vector<int> arr;
        ListNode* temp1 = list1;
        ListNode* temp2 = list2;

        while(temp1 != NULL){
            arr.push_back(temp1->val); 
            temp1 = temp1->next; 
        }

        while(temp2 != NULL){
            arr.push_back(temp2->val);
            temp2 = temp2->next; 
        }

        sort(arr.begin(), arr.end());

        ListNode* dummyNode = new ListNode(-1);
        ListNode* temp = dummyNode;
        for(int i = 0; i < arr.size(); i++){
            temp->next = new ListNode(arr[i]); 
            temp = temp->next; 
        }

        return dummyNode->next; 
    }
};



*/